Index: src/Extra/org/objectweb/proactive/extra/messagerouting/protocol/message/ErrorMessage.java
===================================================================
--- src/Extra/org/objectweb/proactive/extra/messagerouting/protocol/message/ErrorMessage.java	(revision 15006)
+++ src/Extra/org/objectweb/proactive/extra/messagerouting/protocol/message/ErrorMessage.java	(working copy)
@@ -34,7 +34,6 @@
  */
 package org.objectweb.proactive.extra.messagerouting.protocol.message;
 
-import org.objectweb.proactive.core.remoteobject.http.util.HttpMarshaller;
 import org.objectweb.proactive.extra.messagerouting.protocol.AgentID;
 import org.objectweb.proactive.extra.messagerouting.protocol.TypeHelper;
 
@@ -44,6 +43,9 @@
  * Error message are used to notify error to the agent. Several {@link ErrorType} exist.
  * Each one indicate a special type of error.
  * 
+ * Within the current implementation, {@link MessageType#ERR_} is a {@link DataMessage} with the payload
+ * 	being an integer which identifies the error code.
+ * 
  * @since ProActive 4.1.0
  */
 public class ErrorMessage extends DataMessage {
@@ -112,16 +114,44 @@
          */
         ERR_INVALID_ROUTER_ID;
 
-        public byte[] toByteArray() {
-            byte[] buf = new byte[4];
-            TypeHelper.intToByteArray(this.ordinal(), buf, 0);
-            return buf;
+        /** Reverse map associating an error type to an ID  */
+        final static Map<Integer, ErrorType> idToErrorType;
+        static {
+            // Can't populate idToErrorType from constructor since enums are initialized before 
+            // any static initializers are run. It is safe to do it from this static block
+            idToErrorType = new HashMap<Integer, ErrorType>();
+            for (ErrorType errorType : values()) {
+                idToErrorType.put(errorType.ordinal(), errorType);
+            }
         }
+
+        public static ErrorType getErrorType(int value) {
+            return idToErrorType.get(value);
+        }
     }
 
     /** The type of this error message */
     final private ErrorType error;
 
+    /** Read the error type from the payload field of a raw message
+     * It is assumed that the payload of an error message
+     * 	is four bytes long, containing an encoded int
+     *  
+     * @throws IllegalArgumentException - if the payload contains an unrecognized error code
+     * @throws IllegalArgumentException - if the payload is not a four-bytes buffer 
+     * */
+    static public ErrorType readErrorType(byte[] payload) throws IllegalArgumentException {
+        if (payload.length != 4)
+            throw new IllegalArgumentException("The payload is not four-bytes long");
+
+        int errorCode = TypeHelper.byteArrayToInt(payload, 0);
+        ErrorType type = ErrorType.getErrorType(errorCode);
+        if (type != null)
+            return type;
+        else
+            throw new IllegalArgumentException("Invalid value for the error code: " + errorCode);
+    }
+
     /** Create an error message
      *
      * @param recipient 
@@ -134,7 +164,9 @@
      * 		The error type
      */
     public ErrorMessage(ErrorType error, AgentID recipient, AgentID faulty, long msgID) {
-        super(MessageType.ERR_, faulty, recipient, msgID, HttpMarshaller.marshallObject(error));
+        super(MessageType.ERR_, faulty, recipient, msgID, new byte[Integer.SIZE / 8]);
+        // fill in the payload
+        TypeHelper.intToByteArray(error.ordinal(), this.data, 0);
         this.error = error;
     }
 
@@ -154,11 +186,7 @@
             throw new IllegalArgumentException("Invalid message type " + this.getType());
         }
 
-        try {
-            this.error = (ErrorType) HttpMarshaller.unmarshallObject(this.getData());
-        } catch (ClassCastException e) {
-            throw new IllegalArgumentException("Invalid error type:" + e);
-        }
+        this.error = readErrorType(this.getData());
     }
 
     /** Return the type of the error*/
@@ -194,5 +222,10 @@
             return false;
         return true;
     }
+    
+    @Override
+    public String toString() {
+        return super.toString() + " Error code:" + this.error;
+    }
 
 }
