Index: src/Core/org/objectweb/proactive/core/ssh/SSHClient.java
===================================================================
--- src/Core/org/objectweb/proactive/core/ssh/SSHClient.java	(révision 14670)
+++ src/Core/org/objectweb/proactive/core/ssh/SSHClient.java	(copie de travail)
@@ -156,6 +156,7 @@
         }
 
         hostname = remArgs.remove(0);
+        int exitCode = 0;
 
         try {
             Connection conn = new Connection(hostname);
@@ -239,7 +240,7 @@
 
                         /* The remote side won't send us further data... */
                         if ((conditions & (ChannelCondition.STDOUT_DATA | ChannelCondition.STDERR_DATA)) == 0) {
-
+                            exitCode = sess.getExitStatus();
                             /* ... and we have consumed all data in the local arrival window. */
                             break;
                         }
@@ -278,6 +279,6 @@
             e.printStackTrace(System.err);
             System.exit(2);
         }
-        System.exit(0);
+        System.exit(exitCode);
     }
 }
